examples/joke_bot/src/main.rs

```diff
@@ -0,0 +1,268 @@
+                if let Err(err) = write_state(&State {
+                    access_token: client.access_token().expect("Matrix access token is missing"),
+                })
+                .await
+                {
+                    println!("Failed to persist access token to disk. Re-authentication will be required on the next startup: {}", err)
+                }
+                client
+            }
+            Err(e) => {
+                let reason = match e {
+                    client::Error::AuthenticationRequired => {
+                        "invalid credentials specified".to_owned()
+                    }
+                    client::Error::Response(response_err) => {
+                        format!("failed to get a response from the server: {}", response_err)
+                    }
+                    client::Error::FromHttpResponse(parse_err) => {
+                        format!("failed to parse log in response: {}", parse_err)
+                    }
+                    _ => e.to_string(),
+                };
+                panic!("Failed to log in: {}", reason);
+            }
+        }
+    } else {
+        panic!("No previous session found and no credentials stored in config")
+    };
+
+    let filter = FilterDefinition::ignore_all().into();
+    let initial_sync_response = matrix_client
+        .send_request(assign!(sync_events::Request::new(), {
+            filter: Some(&filter),
+        }))
+        .await?;
+    let user_id = &config.username;
+    let not_senders = &[user_id.clone()];
+    let filter = {
+        let mut filter = FilterDefinition::empty();
+        filter.room.timeline.not_senders = not_senders;
+        filter
+    }
+    .into();
+
+    let mut sync_stream = Box::pin(matrix_client.sync(
+        Some(&filter),
+        initial_sync_response.next_batch,
+        &PresenceState::Online,
+        Some(Duration::from_secs(30)),
+    ));
+    println!("Listening...");
+    while let Some(response) = sync_stream.try_next().await? {
+        println!("{}", response.next_batch);
+        for (room_id, room_info) in response.rooms.join {
+            for e in &room_info.timeline.events {
+                match handle_messages(&http_client, &matrix_client, e, &room_id, user_id).await {
+                    Ok(_) => {}
+                    Err(err) => {
+                        println!("failed to respond to message: {}", err)
+                    }
+                }
+            }
+        }
+
+        for (room_id, _) in response.rooms.invite {
+            match handle_invitations(&http_client, &matrix_client, &room_id).await {
+                Ok(_) => {}
+                Err(err) => println!("failed to accept invitation for room {}: {}", &room_id, err),
+            }
+        }
+    }
+    Ok(())
+}
+
+async fn handle_messages(
+    http_client: &HttpClient,
+    matrix_client: &MatrixClient,
+    e: &Raw<AnySyncRoomEvent>,
+    room_id: &RoomId,
+    user_id: &UserId,
```

Comment:
Just for clarity, I'd suggest

```suggestion
    own_user_id: &UserId,
```